File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/c++/PabloGradolph.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <time.h>
using namespace std;

/*
 * Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
 * Podrás configurar generar contraseñas con los siguientes parámetros:
 * - Longitud: Entre 8 y 16.
 * - Con o sin letras mayúsculas.
 * - Con o sin números.
 * - Con o sin símbolos.
 * (Pudiendo combinar todos estos parámetros entre ellos)
 */

string PasswordGenerator(int length, bool upercase, bool numbers, bool symbols){

    // Control de errores en la longitud de la contraseña
    if (length<8 || length>12){
        return "Error en la longitud de la contraseña";
    }

    // Definimos las variables de uso
    string password = "";
    int a;

    // Inicializamos los números aleatorios
    srand(time(NULL));

    // Cuando se acepta todo en la contraseña
    if (upercase && numbers && symbols){
        for (int i=0; i<length; i++){
            a = 33 + rand() % (127 - 33); // Número aleatorio entre 33 y 126 (ASCII)
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos números y mayúsculas
    if (upercase && numbers){
        for (int i=0; i<length; i++){
            a = 48 + rand() % (123 - 48); // Número aleatorio entre 48 y 122 (ASCII)
            while ((a>=58 && a<=64) || (a>=91 && a<=96)){
                a = 48 + rand() % (123 - 48); // Pedimos a hasta que esté en esos números porque sino serían símbolos.
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos números y símbolos
    if (symbols && numbers){
        for (int i=0; i<length; i++){
            a = 33 + rand() % (127 - 33); 
            while (a>=65 && a<=90){
                a = 33 + rand() % (127 - 33);
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos mayúsculas y símbolos
    if (symbols && upercase){
        for (int i=0; i<length; i++){
            a = 33 + rand() % (127 - 33); 
            while (a>=48 && a<=57){
                a = 33 + rand() % (127 - 33);
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos solo mayúsculas.
    if (upercase){
        for (int i=0; i<length; i++){
            a = 65 + rand() % (123 - 65); 
            while (a>=91 && a<=96){
                a = 65 + rand() % (123 - 65); 
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos solo símbolos
    if (symbols){
        for (int i=0; i<length; i++){
            a = 33 + rand() % (127 - 33); 
            while ((a>=48 && a<=57) || (a>=65 && a<=90)){
                a = 33 + rand() % (127 - 33);
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando aceptamos solo números
    if (numbers){
        for (int i=0; i<length; i++){
            a = 48 + rand() % (123 - 48); 
            while (a>=58 && a<=96){
                a = 48 + rand() % (123 - 48);
            }
            password = password + char(a);
        }
        return password;
    }

    // Cuando no aceptamos nada
    if (numbers == false && upercase == false && symbols == false){
        for (int i=0; i<length; i++){
            a = 97 + rand() % (123 - 97); 
            password = password + char(a);
        }
        return password;
    }
}

int main(){
    string password;

    password = PasswordGenerator(8, true, true, true);
    cout<<"Contraseña 1: "<<password<<endl<<endl;
    password = PasswordGenerator(9, true, true, false);
    cout<<"Contraseña 2: "<<password<<endl<<endl;
    password = PasswordGenerator(10, true, false, true);
    cout<<"Contraseña 3: "<<password<<endl<<endl;
    password = PasswordGenerator(11, false, false, true);
    cout<<"Contraseña 4: "<<password<<endl<<endl;
    password = PasswordGenerator(12, false, false, false);
    cout<<"Contraseña 5: "<<password<<endl<<endl;
    password = PasswordGenerator(12, false, true, true);
    cout<<"Contraseña 6: "<<password<<endl<<endl;

    return 0;
}

File: Retos/Reto #3 - EL GENERADOR DE CONTRASEÑAS [Media]/python/PabloGradolph.py
'''
/*
 * Escribe un programa que sea capaz de generar contraseñas de forma aleatoria.
 * Podrás configurar generar contraseñas con los siguientes parámetros:
 * - Longitud: Entre 8 y 16.
 * - Con o sin letras mayúsculas.
 * - Con o sin números.
 * - Con o sin símbolos.
 * (Pudiendo combinar todos estos parámetros entre ellos)
 */
'''

import string, random

def pasword_generator(length: int, uppercase: bool, numbers: bool, symbols: bool) -> str:

    # Definimos los números, letras y signos que podemos tener en la contraseña.
    if uppercase: letters = string.ascii_letters 
    else: letters = string.ascii_lowercase

    if numbers: letters += string.digits 
    if symbols: letters += string.punctuation
    
    # Generamos la contraseña
    pasword = ""
    for _ in range(0, length):
        pasword += random.choice(letters)

    return pasword

def main():

    # Pedimos las características de la contraseña con control de errores.
    length = int(input("Introduzca la longitud deseada de la contraseña: "))
    while length < 8 or length > 16:
        print("La longitud de la contraseña debe estar entre 8 y 16")
        length = int(input("Introduzca de nuevo la longitud de la contraseña: "))

    upper_letters = input("¿Puede tener letras mayúsculas? Y/N: ")
    while upper_letters != "Y" and upper_letters != "N":
        print("Por favor introduzca Y o N (Yes or No)")
        upper_letters = input("¿Puede tener letras mayúsculas? Y/N: ")
    if upper_letters == "Y": upper_letters = True
    else: upper_letters = False

    numbers = input("¿Puede tener números? Y/N: ")
    while numbers != "Y" and numbers != "N":
        print("Por favor introduzca Y o N (Yes or No)")
        numbers = input("¿Puede tener números? Y/N: ")
    if numbers == "Y": numbers = True
    else: numbers = False

    symbols = input("¿Puede tener símbolos? Y/N: ")
    while symbols != "Y" and symbols != "N":
        print("Por favor introduzca Y o N (Yes or No)")
        symbols = input("¿Puede tener símbolos? Y/N: ")
    if symbols == "Y": symbols = True
    else: symbols = False
    
    print(f"Su contraseña es: {pasword_generator(length, upper_letters, numbers, symbols)}")

if __name__ == "__main__":
    main()

